homeassistant/components/songpal/config_flow.py

```diff
@@ -0,0 +1,138 @@
+"""Config flow to configure songpal component."""
+import logging
+from typing import Optional
+from urllib.parse import urlparse
+
+from songpal import Device, SongpalException
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components import ssdp
+from homeassistant.const import CONF_HOST, CONF_NAME
+
+from .const import CONF_ENDPOINT, DOMAIN  # pylint: disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SongpalConfig:
+    """Device Configuration."""
+
+    def __init__(self, name, host, endpoint):
+        """Initialize Configuration."""
+        self.name = name
+        self.host = host
+        self.endpoint = endpoint
+
+
+class SongpalConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Songpal configuration flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize the flow."""
+        self.conf: Optional[SongpalConfig] = None
+
+    async def _show_setup_form(self, user_input=None, errors=None):
+        user_input = user_input or {}
+        default_endpoint = user_input.get(CONF_ENDPOINT)
+        if default_endpoint is not None:
+            data_schema = {vol.Required(CONF_ENDPOINT, default=default_endpoint): str}
+        else:
+            data_schema = {vol.Required(CONF_ENDPOINT): str}
```

Comment:
You do not need all of this complexity. This can be rewritten as:
```python
data_schema = {vol.Required(CONF_ENDPOINT, default=user_input.get(CONF_ENDPOINT)): str}
```
I tested this as follows:
```python
>>> import voluptuous as vol
>>> schema = vol.Schema({vol.Required("test", default=None): str})
>>> schema({})
voluptuous.error.MultipleInvalid: expected str for dictionary value @ data['test']
>>> schema({"test": None})
voluptuous.error.MultipleInvalid: expected str for dictionary value @ data['test']
>>> schema({"test": 1})
voluptuous.error.MultipleInvalid: expected str for dictionary value @ data['test']
```